feat: Add config for attachments download permissions

jan-auer opened the pull request:
This PR adds an organization option to configure who can download event attachments. 

Attachments are potentially sensitive files containing significant PII or user data, such as native crash reports or log files. Some of our customers would like better control over who has access to those attachments. We decided to let users configure the membership role required to access attachment files.

The new option allows to override the default setting of "Member" on an organization level. It is enforced in the download view but does not constrain listing attachments:

<img width="1159" alt="Screenshot 2019-05-15 at 16 01 41" src="https://user-images.githubusercontent.com/1433023/57782015-a9f1b580-772b-11e9-975a-17b642e48f50.png">

Users with insufficient permissions will still see the attachment, but won't be able to download it:

<img width="927" alt="Screenshot 2019-05-15 at 16 00 06" src="https://user-images.githubusercontent.com/1433023/57782056-b8d86800-772b-11e9-9f19-c15662f51fc7.png">

Adding @dcramer for sanity check, @mitsuhiko and @markstory for overall review and @billyvg for UI.
Note that this PR does not conflict with #12995

dcramer commented:
To avoid unneeded complexity, can we just provide this at the organization level?

markstory commented on src/sentry/api/endpoints/event_attachment_details.py:
```diff
@@ -8,12 +8,43 @@
+            current_role = OrganizationMember.objects.filter(
+                organization=organization,
+                user=request.user,
+            ).values_list('role', flat=True).get()
```
Do superusers need access? as they won't have roles on the organization.

markstory commented on src/sentry/api/endpoints/event_attachment_details.py:
```diff
@@ -8,12 +8,43 @@
+        if not result:
+            return result
+
+        if not request.user.is_authenticated():
```
Do organization API keys need access to attachments?

jan-auer replied:
This points me to a pretty serious bug right now, thanks! 
To your question: Are these the keys for integrations or is that yet another thing?

markstory replied:
I was thinking of the older organization API tokens, so it likely isn't that relevant.

jan-auer replied:
Thanks :) Those seem to be deprecated, so I guess we can ignore that 👍

markstory commented on src/sentry/api/endpoints/organization_details.py:
```diff
@@ -141,6 +144,14 @@ def validate_safeFields(self, attrs, source):
+        try:
+            roles.get(value)
+        except KeyError:
+            raise serializers.ValidationError('Invalid role')
```
Should we tell the user what the valid states are?

mitsuhiko replied:
This should never happen anyways unless someone bypasses the UI.

billyvg commented on src/sentry/static/sentry/app/components/events/eventAttachments.jsx:
```diff
@@ -2,18 +2,47 @@ import PropTypes from 'prop-types';
+  const roleIds = availableRoles.map(role => role.id);
+  const requiredIndex = roleIds.indexOf(requiredRole);
+  const currentIndex = roleIds.indexOf(currentRole);
+  return currentIndex >= requiredIndex;
```
Are roles guaranteed to be sorted like this? Is it possible for both required and current indexes to be `-1`?

jan-auer replied:
Correct, roles are sorted by ascending priority (or actually, priority is the index in the list). Only with invalid inputs either could be `-1`, but the API should have validated this before.

jan-auer commented:
Added permission tests for the endpoint and moved UI logic to a reusable place (chose a render-prop component). I chose to hide minidumps in the event table for users who do not have download permissions, as it's pretty useless to show them (and use up space).
